lib/read/lib/Cardano/Wallet/Read/Tx/TxOut.hs

```diff
@@ -0,0 +1,281 @@
+    , Era (..)
+    , EraValue (..)
+    , IsEra (theEra)
+    , indexOfEra
+    , parseEraIndex
+    )
+import Cardano.Wallet.Read.Tx
+    ( Tx (..)
+    )
+import Cardano.Wallet.Read.Tx.TxIn
+    ( TxIn
+    , pattern TxIn
+    , pattern TxIx
+    )
+import Cardano.Wallet.Read.Value
+    ( Value
+    , fromEraValue
+    , toMaryValue
+    )
+import Data.Foldable
+    ( toList
+    )
+import Data.Maybe
+    ( fromMaybe
+    )
+import Data.Maybe.Strict
+    ( StrictMaybe (SNothing)
+    )
+import Data.Word
+    ( Word16
+    )
+
+import qualified Cardano.Ledger.Babbage.TxBody as Babbage
+import qualified Cardano.Wallet.Read.Tx.ScriptValidity as Read
+import qualified Cardano.Wallet.Read.Tx.TxId as Read
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+
+{-----------------------------------------------------------------------------
+    Type
+------------------------------------------------------------------------------}
+-- | A 'TxOut' is a transaction output from any era — past, present
+-- or next one.
+newtype TxOut = TxOutC (EraValue Output)
+
+-- | Test whether two 'TxOut' are equal,
+-- in a way that is independent
+-- of the particular era-dependent internal representation.
+instance Eq TxOut where
+    {-# INLINABLE (==) #-}
+    (TxOutC x) == (TxOutC y) = eqTxOut x y
+
+-- | Make a basic 'TxOut' from an address and a value.
+mkBasicTxOut :: CompactAddr -> Value -> TxOut
+mkBasicTxOut addr value =
+    TxOutC (EraValue (Output txout :: Output Conway))
+  where
+    val = toMaryValue value
+    -- Cardano.Ledger uses the same error message when converting
+    cVal = fromMaybe (error ("Illegal Value in TxOut: " ++ show val))
+        $ toCompact val
+    txout = Babbage.TxOutCompact addr cVal
+
+{-# INLINEABLE getCompactAddr #-}
+-- | Get the address which controls who can spend the transaction output.
+getCompactAddr :: TxOut -> CompactAddr
+getCompactAddr (TxOutC (EraValue txout)) =
+    fromEraCompactAddr $ getEraCompactAddr txout
+
+{-# INLINEABLE getValue #-}
+-- | Get the monetary 'Value' in this transaction output.
+getValue :: TxOut -> Value
+getValue (TxOutC (EraValue txout)) =
+    fromEraValue $ getEraValue txout
+
+-- | Upgrade the internal representation of a 'TxOut' to the 'Conway' era.
+upgradeTxOutToConway :: TxOut -> TxOut
+upgradeTxOutToConway (TxOutC (EraValue txout)) =
+    TxOutC (EraValue (upgradeOutputToConway txout))
+
+-- Test for equality.
+eqTxOut :: EraValue Output -> EraValue Output -> Bool
+eqTxOut (EraValue (v :: Output erax)) (EraValue (w :: Output eray)) =
+    case (theEra :: Era erax, theEra :: Era eray) of
+        (Byron, Byron) -> v == w
+        (Shelley, Shelley) -> v == w
+        (Allegra, Allegra) -> v == w
+        (Mary, Mary) -> v == w
+        (Alonzo, Alonzo) -> v == w
+        (Babbage, Babbage) -> v == w
+        (Conway, Conway) -> v == w
+        (_, _) -> upgradeOutputToConway v == upgradeOutputToConway w
```

Comment:
So, it looks like we have (at least) three potential notions of equality:

1. Equality as induced by `instance Eq a => Eq (EraValue a)` where values from different eras are always considered unequal.
2. Equality as induced by conversion to the next era, i.e. `o == o'` defined as `= toWriteTxOut Conway o == toWriteTxOut Conway o'`
3. Equality as induced by conversion to the current era, i.e.  `o == o'` defined as `= toWriteTxOut Babbage o == toWriteTxOut Babbage o'` when `o` and `o'` are from any era from `Babbage` or earlier. Outputs from Conway are always considered distinct from the previous ones.

> I think a property like this would make sense:
> `forall o o' era. (o == o') == (toWriteTxOut era o == toWriteTxOut era o')`

Yes, this property would make a lot of sense, especially restricted to `era = Conway`. But — it does hold for definition 2.

> but this will fail for the ptr examples above, where Equality is more lenient in Conway than in Babbage.

Under definition 1, this property will fail whenever `o` and `o'` are from different eras, but as you point out, it can also fail if `o` and `o'` are from the same era but involve weird ptr addresses.